tests/unit/httpd_test.cc

```diff
@@ -1218,3 +1219,77 @@ SEASTAR_TEST_CASE(test_shared_future) {
 
     return std::move(fut).discard_result();
 }
+
+static auto graceful_stop_tester(bool blocking) {
+    return seastar::async([blocking] {
+        loopback_connection_factory lcf;
+        http_server server("test");
+        loopback_socket_impl lsi(lcf);
+        httpd::http_server_tester::listeners(server).emplace_back(lcf.get_server_socket());
+
+        auto handler = new echo_string_handler();
+        server._routes.put(GET, "/test", handler);
+        server.do_accepts(0).get();
+
+        promise<> client_block;
+
+        auto client = async([&lsi, &client_block] () {
+            connected_socket c_socket = lsi.connect(socket_address(ipv4_addr()), socket_address(ipv4_addr())).get0();
+            input_stream<char> input(c_socket.input());
+            output_stream<char> output(c_socket.output());
+
+            // TODO(nikandfor): should be here, but breaks test with exception "error system:32, Broken pipe" and failed assert
+            // include/seastar/core/gate.hh:56: seastar::gate::~gate(): Assertion `!_count && "gate destroyed with outstanding requests"' failed.
```

Comment:
`"gate destroyed with outstanding requests"` suggests that somewhere in the code nobody waited for `gate.close()` to finish and the gate was destroyed, which is a bug.

Most likely your problem can be fixed with `seastar::defer`, to make sure that you call `server::stop()`, even if the function exits early due to an exception. I didn't really follow the whole logic of this test case, but perhaps instead of placing
```cpp
server.stop().get();
```
 at the end of it, you should create a deferred action right after initializing the server:
```cpp
auto stop_server = seastar::defer([&server] {
    server.stop().get();
});
```
Then, even if the function exits early and unrolls the stack, the server will be stopped properly, waiting for the gate to be closed and for all the outstanding requests to finish.

Reply:
Thank you, that worked.